docs/source/configure.rst
.. _configuration_toplevel:

.. _configuring_stalker:

Configuring Stalker
===================

To configure Stalker and make it fit to your Studios need you should use the
``config.py`` file as mentioned in next sections.

config.py File
--------------

Stalker uses the ``config.py`` to let one to customize the system config.

The ``config.py`` file is searched in a couple of places through the system:
    
  * under "~/.strc/" directory (not yet)
  * under "$STALKER_PATH"

The first path is a folder in the users home dir. The second one is a path
defined by the ``STALKER_PATH`` environment variable.

Defining the ``config.py`` by using the environment variable gives the most
customizable and consistent setup through the studio. You can set
``STALKER_PATH`` to a shared folder in your fileserver where all the users can
access.

Because, ``config.py`` is a regular Python code which is executed by
Stalker, you can do anything you were doing in a normal Python
script. This is very handy (also dangerous!) if you have another source of
information which is reachable by a Python script.

If there is no ``STALKER_PATH`` variable in your current environment or it is
not showing an existing path or there is no ``config.py`` file the system will
use the system defaults.

Config Variables
----------------

Variables which can be set in ``config.py`` are as follows:

.. confval:: actions

   Actions for authorization system. These are used to create ACLs. Stalker
   uses `CRUDL`_ system. Default value is::

     actions = ['Create', 'Read', 'Update', 'Delete', 'List'] #CRUDL

   .. _CRUDL: http://en.wikipedia.org/wiki/Create,_read,_update_and_delete

.. confval:: auto_create_admin

   Tells Stalker to create an admin by default. Default value is::

     auto_create_admin = True

.. confval:: admin_name

   The default admin user name. Default value is::

     admin_name = 'admin'

.. confval:: admin_login

   The default admin login. Default value is::

     admin_login = 'admin'

.. confval:: admin_password

   The default admin password. Default value is::

     admin_password = 'admin'

.. confval:: admin_email

   The default email for admin user. Default value is::

     admin_email = 'admin@admin.com'

.. confval:: admin_department_name

   The default department name for admin. Default value is::

     admin_department_name = 'admins'

.. confval:: admin_group_name

   The default admin permission group name. Default value is::

     admin_group_name = 'admins'

.. confval:: database_engine_settings

   A dictionary of config values. The default value is::

     database_engine_settings = {
         "sqlalchemy.url": "sqlite:///:memory:",
         "sqlalchemy.echo": False,
     }

.. confval:: database_session_settings
   
   This value is not used.

.. confval:: local_storage_path

   The local storage path for Stalker.

     local_storage_path = os.path.expanduser('~/.strc')   

.. confval:: local_session_data_file_name

   The per user or local session file name. It is used for storing logged in
   user info. The default value is::

     local_session_data_file_name = 'local_session_data'

.. confval:: server_side_storage_path

   Storage for uploaded files. This used by `Stalker Pyramid`_ and shows the
   server side storage path. Will be moved to Stalker Pyramid in later
   versions. Not used by Stalker by default. Default value is::

     server_side_storage_path = os.path.expanduser('~/Stalker_Storage')

   .. _`Stalker Pyramid`: https://pypi.python.org/pypi/stalker_pyramid

.. confval:: key

   The default keyword which is going to be used in password scrambling.
   Default value is::

     key = "stalker_default_key"

.. confval:: version_variant_name

   The default variant name for :class:`~stalker.models.version.Version`
   instances. Default value is::

     version_variant_name = "Main"

.. confval:: status_bg_color

   Default background color for :class:`~stalker.models.status.Status`
   instances. Default value is::

     status_bg_color = 0xffffff

.. confval:: status_fg_color

   Default foreground color for :class:`~stalker.models.status.Status`
   instances. Default value is::

     status_fg_color = 0x000000
 
.. confval:: ticket_label

   Default ticket label. Used by :class:`~stalker.models.ticket.Ticket` when
   generating a ticket name. Default value is::

     ticket_label = "Ticket"

.. confval:: ticket_status_order

   Defines the ticket statuses and the order of them. Default value is::

     ticket_status_order = [
         'new', 'accepted', 'assigned', 'reopened', 'closed'
     ]

.. confval:: ticket_resolutions

   Defines the default ticket resolutions. Default value is::

     ticket_resolutions = [
         'fixed', 'invalid', 'wontfix', 'duplicate', 'worksforme', 'cantfix'
     ]

.. confval:: ticket_workflow

   Defines the default ticket workflow. It is a dictionary of actions. Shows
   the new status per action. Default value is::

     ticket_workflow = {
         'resolve' : {
             'new': {
                 'new_status': 'closed',
                 'action': 'set_resolution'
             },
             'accepted': {
                 'new_status': 'closed',
                 'action': 'set_resolution'
             },
             'assigned': {
                 'new_status': 'closed',
                 'action': 'set_resolution'
             },
             'reopened': {
                 'new_status': 'closed',
                 'action': 'set_resolution'
             },
         },
         'accept' : {
             'new': {
                 'new_status': 'accepted',
                 'action': 'set_owner'
             },
             'accepted': {
                 'new_status': 'accepted',
                 'action': 'set_owner'
             },
             'assigned': {
                 'new_status': 'accepted',
                 'action': 'set_owner'
             },
             'reopened': {
                 'new_status': 'accepted',
                 'action': 'set_owner'
             },
         },
         'reassign': {
             'new': {
                 'new_status': 'assigned',
                 'action': 'set_owner'
             },
             'accepted': {
                 'new_status': 'assigned',
                 'action': 'set_owner'
             },
             'assigned': {
                 'new_status': 'assigned',
                 'action': 'set_owner'
             },
             'reopened': {
                 'new_status': 'assigned',
                 'action': 'set_owner'
             },
         },
         'reopen': {
             'closed': {
                 'new_status': 'reopened',
                 'action': 'del_resolution'
             }
         }
     }

.. confval:: timing_resolution

   Defines the default timing resolution for classes which are mixed with
   :class:`~stalker.models.mixins.DateRangeMixin`\ . Stalker uses the
   TaskJuggler default timing resolution which is 1 hour::

     timing_resolution = datetime.timedelta(hours=1)

.. confval:: task_duration

   Defines the default task duration. If only a start or end value is entered
   for a :class:`~stalker.models.task.Task` then Stalker calculates the other
   value by adding or subtracting the default task duration value from it.
   Default value is 1 hour::

     task_duration = datetime.timedelta(hours=1)


.. confval:: task_priority

   Defines the default task priority. This is used by TaskJuggler to prioritize
   tasks. Should be a number between 0 and 1000. Default value is 500::

     task_priority = 500

.. confval:: working_hours

   Defines the default weekly working hours per week day. Stalker uses the
   TaskJuggler default value of 9am to 6pm. The values entered are minutes from
   midnight, and it is a list of lists of two integers. Each list of two
   integers shows a working hour interval. Default value is::

     working_hours = {
       'mon': [[540, 1080]], # 9:00 - 18:00
       'tue': [[540, 1080]], # 9:00 - 18:00
       'wed': [[540, 1080]], # 9:00 - 18:00
       'thu': [[540, 1080]], # 9:00 - 18:00
       'fri': [[540, 1080]], # 9:00 - 18:00
       'sat': [], # saturday off
       'sun': [], # sunday off
     }

.. confval:: daily_working_hours

   Defines the default daily working hour. This is strongly related with the
   ``working_hours``, ``weekly_working_hours``, ``weekly_working_days`` and
   ``yearly_working_days`` settings and shows a mean value of daily working
   hour. Default value is 9::

     daily_working_hours = 9

.. confval:: weekly_working_hours

   Defines the default weekly working hour. This is strongly related with the
   ``working_hours``, ``daily_working_hours``, ``weekly_working_days`` and
   ``yearly_working_days`` settings. Default value is 45::

     weekly_working_hours = 45

.. confval:: weekly_working_days

   Defines the default weekly working days. This is strongly related with the
   ``working_hours``, ``daily_working_hours``, ``weekly_working_hours`` and
   ``yearly_working_days`` settings. Default value is 5::

     weekly_working_days = 5

.. confval:: yearly_working_days

   Defines the default yearly working days. This is strongly related with the
   ``working_hours``, ``daily_working_hours``, ``weekly_working_hours`` and
   ``weekly_working_days`` settings. Default value is 260.714 which equals
   ``weekly_working_days`` * 52.1428::

     yearly_working_days = 260.714

.. confval:: day_order

   Defines the order of the week days. Default value uses European system::

     day_order = ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun']

.. confval:: datetime_units

   Defines the date and time units. The order should match the
   ``datetime_unit_names`` setting. Default value is::

     datetime_units = ['min', 'h', 'd', 'w', 'm', 'y']

.. confval:: datetime_unit_names

   Defines the names of date and time units. The order should match the
   ``datetime_units`` setting. Default value is::

     datetime_unit_names = ['minute', 'hour', 'day', 'week', 'month', 'year']

.. confval:: datetime_units_to_timedelta_kwargs

   Defines the conversion ratios of each date and time unit. Default value is::

     datetime_units_to_timedelta_kwargs = {
         'min': {'name': 'minutes', 'multiplier': 1},
         'h'  : {'name': 'hours'  , 'multiplier': 1},
         'd'  : {'name': 'days'   , 'multiplier': 1},
         'w'  : {'name': 'weeks'  , 'multiplier': 1},
         'm'  : {'name': 'days'   , 'multiplier': 30},
         'y'  : {'name': 'days'   , 'multiplier': 365}
     }

.. confval:: task_schedule_models

   Defines the default schedule models. These are highly related with
   TaskJuggler, so anything entered here should exist in TaskJuggler. Default
   value is::

     task_schedule_models = ['effort', 'length', 'duration']

.. confval:: task_schedule_constraints

   Defines the default schedule constraints. The order also defines a binary
   number corresponding to each value (00: none, 01: start, 10:end, 11:both)
   and used in defining which side of a Task is constrained to a date. Also
   used by TaskJuggler to constrain the start or end or both dates of a task to
   a certain date. Also a Task with schedule_constraint is set to 2 (both) is
   considered a **duration** task even if its schedule_model is set to
   **effort** or **length**. Default value is::

     task_schedule_constraints = ['none', 'start', 'end', 'both']

.. confval:: tjp_working_hours_template

   Defines a Jinja2 template for converting
   :class:`~stalker.models.studio.WorkingHours` instances to a TaskJuggler
   compatible string. By default Stalker converts a WorkingHours instance to a
   ``workinghours`` statement in TaskJuggler. Default value is::

     tjp_working_hours_template = """{% macro wh(wh, day) -%}
     {%- if wh[day]|length %}    workinghours {{day}} {% for part in wh[day] -%}
             {%- if loop.index != 1%}, {% endif -%}
             {{"%02d"|format(part[0]//60)}}:{{"%02d"|format(part[0]%60)}} - {{"%02d"|format(part[1]//60)}}:{{"%02d"|format(part[1]%60)}}
             {%- endfor -%}
     {%- else %}    workinghours {{day}} off
     {%- endif -%}
     {%- endmacro -%}
     {{wh(workinghours, 'mon')}}
     {{wh(workinghours, 'tue')}}
     {{wh(workinghours, 'wed')}}
     {{wh(workinghours, 'thu')}}
     {{wh(workinghours, 'fri')}}

CHANGELOG
===============
Stalker Changes
===============

0.2.21
======

* **New:** Switched from ``nose`` + ``unittest`` to ``pytest`` as the main
  testing framework (with ``pytest-xdist`` tests are 4x faster).

* **New:** Added ``DBSession.save()`` shortcut method for convenience which
  does an ``add`` or ``add_all`` (depending to the input) followed by a
  ``commit`` at once.

0.2.20
======

* **New:** Added ``goods`` attribute to the ``Client`` class. To allow special
  priced ``Goods`` to be created for individual clients.

* **Fix:** The ``WorkingHours`` class is now derived from ``Entity`` thus it is
  not stored in a ``PickleType`` column in ``Studio`` anymore. (issue: #44)

* **Update:** Updated ``appveyor.yml`` to match ``travis.yml``.

0.2.19
======

* **Update:** Updated the ``stalker.config.Config.database_engine_settings`` to
  point the test database.

* **Fix:** Fixed a bug in ``stalker.testing.UnitTestDBBase.setUp()`` where it
  was not considering the existence of the ``STALKER_PATH`` environment
  variable while doing the tests.

* **Update:** Removed debug message from ``db.setup()`` which was revealing the
  database password.

* **Update:** Updated the ``UnitTestDBBase``, it now creates its own test
  database, which allows all the tests to run in an individual database. Thus,
  the tests can now be run in ``multiprocess`` mode which speeds things a lot.

* **Fix:** Removed any module level imports of ``stalker.defaults`` variable,
  which can be changed by a Studio (or by tests) and should always be
  refreshed.

* **Update:** Removed the module level import of the
  ``stalker.db.session.DBSession`` in ``stalker.db``, so it is not possible to
  use ``db.DBSession`` anymore.

* **Update:** The import statements that imports ``stalker.defaults`` moved to
  local scopes to allow runtime changes to the ``defaults`` to be reflected
  correctly.

* **Update:** Added Python fall back mode to
  ``stalker.shot.Shot._check_code_availability()`` which runs when there is no
  database.

* **Update:** ``stalker.models.task.TimeLog._validate_task()`` is now getting
  the ``Status`` instances from the ``StatusList`` that is attached to the
  ``Task`` instance instead of doing a database query.

* **Update:** ``stalker.models.task.TimeLog._validate_resource()`` is now
  falling back to a Python implementation if there is no database connection.

* **Update:** ``stalker.models.task.Task._total_logged_seconds_getter()`` is
  now hundreds of times faster when there is a lot of ``TimeLog`` instances
  attached to the ``Task``.

* **Update:** In ``stalker.models.task.Task`` class, methods those were doing a
  database query to get the required ``Status`` instances are now using the
  attached ``StatusList`` instance to get them.

* **Fix:** A possible ``auto_flush`` is prevented in ``Ticket`` class.

* **Update:** ``Version.latest_version`` property is now able to fall back to a
  pure Python implementation when there is no database connection.

* **Update:** The default log level has been increased from ``DEBUG`` to
  ``INFO``.

* **Update:** In an attempt to speed up tests, a lot of tests that doesn't need
  an active Database has been updated to use the regular ``unittest.TestCase``
  instead of ``stalker.testing.TestBase`` and as a result running all of the
  tests are now 2x faster.

* **Fix:** ``TimeLogs`` are now correctly reflected in UTC in a tj3 file.

* **Fix:** Fixed a lot of tests which were raising Warnings and surprisingly
  considered as Errors in TravisCI.

* **Fix:** ``to_tjp`` methods of SOM classes that is printing a Datetime object
  are now printing the dates in UTC.

* **Fix:** Fixed ``stalker.models.auth.Permission`` to be hashable for Python
  3.

* **Fix:** Fixed ``stalker.models.auth.AuthenticationLog`` to be sortable for
  Python 3.

* **Fix:** Fixed ``stalker.models.version.Version.latest_version`` property for
  Python 3.

* **Fix:** Fixed tests of ``Permission`` class to check for correct exception
  messages in Python 3.

* **Update:** Replaced the ``assertEquals`` and ``assertNotEquals`` calls which
  are deprecated in Python 3 with ``assertEqual`` and ``assertNotEquals`` calls
  respectively.

* **Fix:** Fixed tests for ``User`` and ``Version`` classes to not to cause the
  ``id column is None`` warnings of SQLAlchemy to be emitted.


0.2.18
======

* **Update:** Support for DB backends other than Postgresql has been dropped.
  This is done to greatly benefit from a code that is highly optimized only
  for one DB backend. With This all of the tests should be inherited from the
  ``stalker.tests.UnitTestDBBase`` class.

* **New:** All the DateTime fields in Stalker are now TimeZone aware and
  Stalker stores the DateTime values in UTC. Naive datetime values are not
  supported anymore. You should use a library like ``pytz`` to supply timezone
  information as shown below::

    import datetime
    import pytz
    from stalker import db, SimpleEntity
    new_simple_entity = SimpleEntity(
        name='New Simple Entity',
        date_created = datetime.datetime.now(tzinfo=pytz.utc)
    )

* **Fix:** The default values for ``date_created`` and ``date_updated`` has now
  been properly set to a partial function that returns the current time.

* **Fix:** Previously it was possible to enter two TimeLogs for the same
  resource in the same datetime range by committing the data from two different
  sessions simultaneously. Thus the database was not aware that it should
  prevent that. Now with the new PostgreSQL only implementation and the
  ``ExcludeConstraint`` of PostgreSQL an ``IntegrityError`` is raised by the
  database backend when something like that happens.

* **Update:** All the tests those are checking the system against an Exception
  is being raised or not are now checking also the exception message.

* **Update:** In the ``TimeLog`` class, the raised ``OverBookedException``
  message has now been made clear by adding the start and end date values of

src/stalker/config.py
# -*- coding: utf-8 -*-
"""Config related functions and classes are situated here."""
import datetime
import os
import sys
from typing import Any

from stalker import log

logger = log.get_logger(__name__)


class ConfigBase(object):
    """Config abstraction.

    This is based on Sphinx's config idiom.
    """

    default_config_values = {}

    def __init__(self):
        self.config_values = self.default_config_values.copy()
        self.user_config = {}
        self._parse_settings()

    def _parse_settings(self) -> None:
        """Parse the settings.

        The priority order is:

            stalker.config
            config.py under .stalker_rc directory
            config.py under $STALKER_PATH

        Raises:
            RuntimeError: If there is a Syntax error in the configuration.
        """
        # for now just use $STALKER_PATH
        # try to get the environment variable
        if self.env_key not in os.environ:
            # don't do anything
            logger.debug("no environment key found for user settings")
        else:
            logger.debug("environment key found")

            resolved_path = os.path.expanduser(
                os.path.join(os.environ[self.env_key], "config.py")
            )

            # using `while` is not safe to expand variables
            # so expand vars for 100 times which already is ridiculously
            # complex
            max_recursion = 100
            i = 0
            while "$" in resolved_path and i < max_recursion:
                resolved_path = os.path.expandvars(resolved_path)
                i += 1

            try:
                logger.debug("importing user config")
                with open(resolved_path) as f:
                    exec(f.read(), self.user_config)
            except IOError:
                logger.warning(
                    f"The $STALKER_PATH: {resolved_path} doesn't exists! "
                    "skipping user config"
                )
            except SyntaxError as e:
                raise RuntimeError(
                    f"There is a syntax error in your configuration file: {e}"
                )
            finally:
                # append the data to the current settings
                logger.debug("updating system config")
                for key in self.user_config:
                    # if key in self.config_values:
                    self.config_values[key] = self.user_config[key]

    def __getattr__(self, name: str) -> Any:
        """Return the config value as if it is an attribute look up.

        Args:
            name (str): The name of the config value.

        Returns:
            Any: The value related to the given config value.
        """
        return self.config_values[name]

    def __getitem__(self, name) -> Any:
        """Return item with the key.

        Args:
            name (str): The key to find the value of.

        Returns:
            Any: The value related to the given key.
        """
        return getattr(self, name)

    def __setitem__(self, name, value) -> None:
        """Set the item with index of name to value.

        Args:
            name (str): The name as the index.
            value (Any): The value to set the item to.
        """
        self.config_values[name] = value

    def __delitem__(self, name: str) -> None:
        """Delete the item with the given name.

        Args:
            name (str): The name of the item to delete.
        """
        self.config_values.pop(name)

    def __contains__(self, name: str) -> bool:
        """Check if this contains the name.

        Args:
            name (str): The config name.

        Returns:
            bool: True if this contains the name, False otherwise.
        """
        return name in self.config_values


class Config(ConfigBase):
    """Holds system-wide configuration variables.

    See `configuring stalker`_ for more detail.

    .. _configuring stalker: ../configure.html
    """

    env_key = "STALKER_PATH"

    default_config_values = dict(
        #
        # The default settings for the database, see sqlalchemy.create_engine
        # for possible parameters
        #
        database_engine_settings={
            "sqlalchemy.url": "sqlite://",
            "sqlalchemy.echo": False,
            # "sqlalchemy.pool_pre_ping": True,
        },
        database_session_settings={},
        # Local storage path
        local_storage_path=os.path.expanduser("~/.strc"),
        local_session_data_file_name="local_session_data",
        # Storage for uploaded files
        server_side_storage_path=os.path.expanduser("~/Stalker_Storage"),
        repo_env_var_template="REPO{code}",
        #
        # Tells Stalker to create an admin by default
        #
        auto_create_admin=True,
        #
        # these are for new projects
        # after creating the project you can change them from the interface
        #
        admin_name="admin",
        admin_login="admin",
        admin_password="admin",
        admin_email="admin@admin.com",
        admin_department_name="admins",
        admin_group_name="admins",
        # the default keyword which is going to be used in password scrambling
        key="stalker_default_key",
        version_variant_name="Main",
        actions=["Create", "Read", "Update", "Delete", "List"],  # CRUDL
        # Tickets
        ticket_label="Ticket",
        # define the available actions per Status
        ticket_status_names=["New", "Accepted", "Assigned", "Reopened", "Closed"],
        ticket_status_codes=["NEW", "ACP", "ASG", "ROP", "CLS"],
        ticket_resolutions=[
            "fixed",
            "invalid",
            "wontfix",
            "duplicate",
            "worksforme",
            "cantfix",
        ],
        ticket_workflow={
            "resolve": {
                "New": {"new_status": "Closed", "action": "set_resolution"},
                "Accepted": {"new_status": "Closed", "action": "set_resolution"},
                "Assigned": {"new_status": "Closed", "action": "set_resolution"},
                "Reopened": {"new_status": "Closed", "action": "set_resolution"},
            },
            "accept": {
                "New": {"new_status": "Accepted", "action": "set_owner"},
                "Accepted": {"new_status": "Accepted", "action": "set_owner"},
                "Assigned": {"new_status": "Accepted", "action": "set_owner"},
                "Reopened": {"new_status": "Accepted", "action": "set_owner"},
            },
            "reassign": {
                "New": {"new_status": "Assigned", "action": "set_owner"},
                "Accepted": {"new_status": "Assigned", "action": "set_owner"},
                "Assigned": {"new_status": "Assigned", "action": "set_owner"},
                "Reopened": {"new_status": "Assigned", "action": "set_owner"},
            },
            "reopen": {
                "Closed": {"new_status": "Reopened", "action": "del_resolution"}
            },
        },
        # Task Management
        timing_resolution=datetime.timedelta(hours=1),
        task_priority=500,
        working_hours={
            "mon": [[540, 1080]],  # 9:00 - 18:00
            "tue": [[540, 1080]],  # 9:00 - 18:00
            "wed": [[540, 1080]],  # 9:00 - 18:00
            "thu": [[540, 1080]],  # 9:00 - 18:00
            "fri": [[540, 1080]],  # 9:00 - 18:00
            "sat": [],  # saturday off
            "sun": [],  # sunday off
        },
        # this is strongly related with the working_hours settings,
        # this should match each other
        daily_working_hours=9,
        weekly_working_days=5,
        weekly_working_hours=45,
        yearly_working_days=261,  # math.ceil(5 * 52.1428)
        day_order=["mon", "tue", "wed", "thu", "fri", "sat", "sun"],
        datetime_units=["min", "h", "d", "w", "m", "y"],
        datetime_unit_names=["minute", "hour", "day", "week", "month", "year"],
        datetime_units_to_timedelta_kwargs={
            "min": {"name": "minutes", "multiplier": 1},
            "h": {"name": "hours", "multiplier": 1},
            "d": {"name": "days", "multiplier": 1},
            "w": {"name": "weeks", "multiplier": 1},
            "m": {"name": "days", "multiplier": 30},
            "y": {"name": "days", "multiplier": 365},
        },
        task_status_names=[
            "Waiting For Dependency",
            "Ready To Start",
            "Work In Progress",
            "Pending Review",
            "Has Revision",
            "Dependency Has Revision",
            "On Hold",
            "Stopped",
            "Completed",
        ],
        task_status_codes=[
            "WFD",
            "RTS",
            "WIP",
            "PREV",
            "HREV",
            "DREV",
            "OH",
            "STOP",
            "CMPL",
        ],
        project_status_names=["Ready To Start", "Work In Progress", "Completed"],
        project_status_codes=["RTS", "WIP", "CMPL"],
        review_status_names=["New", "Requested Revision", "Approved"],
        review_status_codes=["NEW", "RREV", "APP"],
        daily_status_names=["Open", "Closed"],
        daily_status_codes=["OPEN", "CLS"],
        task_schedule_constraints=["none", "start", "end", "both"],
        task_schedule_models=["effort", "length", "duration"],
        task_dependency_gap_models=["length", "duration"],
        task_dependency_targets=["onend", "onstart"],
        allocation_strategy=[
            "minallocated",
            "maxloaded",
            "minloaded",
            "order",
            "random",
        ],
        persistent_allocation=True,
        tjp_working_hours_template="""
{%- macro wh(wh, day) -%}
{%- if wh[day]|length %}    workinghours {{day}} {% for part in wh[day] -%}
    {%- if loop.index != 1%}, {% endif -%}
        {{"%02d"|format(part[0]//60)}}:{{"%02d"|format(part[0]%60)}} - {{"%02d"|format(part[1]//60)}}:{{"%02d"|format(part[1]%60)}}
    {%- endfor -%}
{%- else %}    workinghours {{day}} off
{%- endif -%}
{%- endmacro -%}
{{wh(workinghours, 'mon')}}
{{wh(workinghours, 'tue')}}
{{wh(workinghours, 'wed')}}
{{wh(workinghours, 'thu')}}
{{wh(workinghours, 'fri')}}
{{wh(workinghours, 'sat')}}
{{wh(workinghours, 'sun')}}""",  # noqa: B950
        tjp_studio_template="""
project {{ studio.tjp_id }} "{{ studio.tjp_id }}" {{ studio.start.date() }} - {{ studio.end.date() }} {
    timingresolution {{ '%i'|format((studio.timing_resolution.days * 86400 + studio.timing_resolution.seconds)//60|int) }}min
    now {{ studio.now.astimezone(utc).strftime('%Y-%m-%d-%H:%M') }}
    dailyworkinghours {{ studio.daily_working_hours }}
    weekstartsmonday
{{ studio.working_hours.to_tjp }}
    timeformat "%Y-%m-%d"
    scenario plan "Plan"
    trackingscenario plan
}
""",  # noqa: B950
        tjp_project_template="""
task {{project.tjp_id}} "{{project.tjp_id}}" {
    {% for task in project.root_tasks %}
        {{task.to_tjp}}
    {%- endfor %}
}
""",
        tjp_task_template="""
task {{task.tjp_id}} "{{task.tjp_id}}" {

    {% if task.priority != 500 -%}
        priority {{task.priority}}
    {%- endif %}

    {% if task.task_depends_on %}
        depends {# #}
        {%- for depends_on in task.task_depends_on %}
            {%- if loop.index != 1 %}, {% endif %}{{depends_on.to_tjp}}
        {%- endfor %}
    {%- endif -%}

    {%- if task.is_container -%}
        {% for child_task in task.children -%}
            {{ child_task.to_tjp }}
        {%- endfor %}
    {%- else %}
        {% if task.resources|length %}

            {% if task.schedule_constraint %}
                {% if task.schedule_constraint == 1 or task.schedule_constraint == 3 %}
                    start {{ task.start.astimezone(utc).strftime('%Y-%m-%d-%H:%M') }}
                {% endif %}
                {% if task.schedule_constraint == 2 or task.schedule_constraint == 3 -%}
                    end {{ task.end.astimezone(utc).strftime('%Y-%m-%d-%H:%M') }}
                {% endif %}
            {% endif %}

            {{task.schedule_model}} {{task.schedule_timing}}{{task.schedule_unit}}
            allocate {# #}
            {%- for resource in task.resources|sort(attribute='id') -%}
                {%- if loop.index != 1 %}, {% endif %}{{ resource.tjp_id }} {# #}
                {%- if task.alternative_resources %}{
                    alternative
                    {% for alt_res in task.alternative_resources -%}
                        {%-if loop.index != 1 %}, {% endif %}
                        {{- alt_res.tjp_id}}
                    {%- endfor -%}
                    {# #} select {{task.allocation_strategy}}
                    {% if task.persistent_allocation -%}
                        persistent
{# #}
                    {%- endif %}
                }
                {%- endif %}
            {%- endfor %}
        {%- endif %}
        {% for time_log in task.time_logs %}
            booking {{time_log.resource.tjp_id}} {{time_log.start.astimezone(utc).strftime('%Y-%m-%d-%H:%M:%S')}} - {{time_log.end.astimezone(utc).strftime('%Y-%m-%d-%H:%M:%S')}} { overtime 2 }
        {% endfor %}
    {% endif %}

}
""",  # noqa: B950
        tjp_task_dependency_template="""{{depends_on.tjp_abs_id}} { {{- dependency_target}}{%if gap_timing %} gap{{gap_model}} {{gap_timing}}{{gap_unit -}}{%endif -%}}""",  # noqa: B950
        tjp_department_template="""
resource {{department.tjp_id}} "{{department.tjp_id}}" {
{% for resource in department.users %}
    {{resource.to_tjp}}
{% endfor -%}
}
""",
        tjp_vacation_template="""vacation {{ vacation.start.astimezone(utc).strftime('%Y-%m-%d-%H:%M:%S') }} - {{ vacation.end.astimezone(utc).strftime('%Y-%m-%d-%H:%M:%S') }}""",  # noqa: B950
        tjp_user_template="""resource {{user.tjp_id}} "{{user.tjp_id}}" {
    efficiency {{user.efficiency}}
{% if user.vacations %}
{% for vacation in user.vacations %}
    {{vacation.to_tjp}}
{% endfor %}
{% endif -%} }""",
        tjp_main_template="""# Generated By Stalker v{{stalker.__version__}}
        {{studio.to_tjp}}

        # resources
        resource resources "Resources" {
        {%- for vacation in studio.vacations %}
            {{vacation.to_tjp}}
        {%- endfor %}
        {%- for user in studio.users %}
            {{user.to_tjp}}
        {%- endfor %}
        }

        # tasks
        {% for project in projects %}
            {{project.to_tjp}}
        {% endfor %}

        # reports
        taskreport breakdown "{{csv_file_name}}"{
            formats csv
            timeformat "%Y-%m-%d-%H:%M"
            columns id, start, end {%- if compute_resources %}, resources{% endif %}
        }
        """,
        tjp_main_template2="""# Generated By Stalker v{{stalker.__version__}}
        {{studio.to_tjp}}

        # resources
        resource resources "Resources" {
        {%- for vacation in studio.vacations %}
            {{vacation.to_tjp}}
        {%- endfor %}
        {%- for user in studio.users %}
            {{user.to_tjp}}
        {%- endfor %}
        }

# tasks
{{ tasks_buffer }}

# reports
taskreport breakdown "{{csv_file_name}}"{
    formats csv
    timeformat "%Y-%m-%d-%H:%M"
    columns id, start, end {%- if compute_resources %}, resources{% endif %}
}""",
        tj_command="tj3" if sys.platform == "win32" else "/usr/local/bin/tj3",
        path_template="{{project.code}}/{%- for parent_task in parent_tasks -%}{{parent_task.nice_name}}/{%- endfor -%}",  # noqa: B950
        filename_template='{{task.entity_type}}_{{task.id}}_{{version.variant_name}}_v{{"%03d"|format(version.version_number)}}',  # noqa: B950
        # --------------------------------------------
        # the following settings came from oyProjectManager
        sequence_format="%h%p%t %R",
        file_size_format="%.2f MB",
        date_time_format="%Y.%m.%d %H:%M",
        resolution_presets={
            "PC Video": [640, 480, 1.0],
            "NTSC": [720, 486, 0.91],
            "NTSC 16:9": [720, 486, 1.21],
            "PAL": [720, 576, 1.067],
            "PAL 16:9": [720, 576, 1.46],
            "HD 720": [1280, 720, 1.0],
            "HD 1080": [1920, 1080, 1.0],
            "1K Super 35": [1024, 778, 1.0],
            "2K Super 35": [2048, 1556, 1.0],
            "4K Super 35": [4096, 3112, 1.0],
            "A4 Portrait": [2480, 3508, 1.0],
            "A4 Landscape": [3508, 2480, 1.0],
            "A3 Portrait": [3508, 4960, 1.0],
            "A3 Landscape": [4960, 3508, 1.0],
            "A2 Portrait": [4960, 7016, 1.0],
            "A2 Landscape": [7016, 4960, 1.0],
            "50x70cm Poster Portrait": [5905, 8268, 1.0],
            "50x70cm Poster Landscape": [8268, 5905, 1.0],
            "70x100cm Poster Portrait": [8268, 11810, 1.0],
            "70x100cm Poster Landscape": [11810, 8268, 1.0],
            "1k Square": [1024, 1024, 1.0],
            "2k Square": [2048, 2048, 1.0],
            "3k Square": [3072, 3072, 1.0],
            "4k Square": [4096, 4096, 1.0],
        },
        default_resolution_preset="HD 1080",
        project_structure="""{% for shot in project.shots %}
                Shots/{{shot.code}}
                Shots/{{shot.code}}/Plate
                Shots/{{shot.code}}/Reference
                Shots/{{shot.code}}/Texture
            {% endfor %}
        {% for asset in project.assets%}
            {% set asset_path = project.full_path + '/Assets/' + asset.type.name + '/' + asset.code %}
            {{asset_path}}/Texture
            {{asset_path}}/Reference
        {% endfor %}
        """,  # noqa: B950
        thumbnail_format="jpg",
        thumbnail_quality=70,
        thumbnail_size=[320, 180],
    )
